windows-containers/productization.md

```diff
@@ -0,0 +1,187 @@
+* Parse the worker ignition and extract the bootstrap kubeconfig and the
+kubelet configuration
+  * We are not using the ignition file during the node booting stage
+* Launch the kubelet as a Windows service
+* Check if the kubelet is running
+* Exit
+
+### Risks and Mitigations
+
+The main risk with this proposal is the dependency on Microsoft to publish a
+downstream version of the kubelet. If this does not happen we would have to
+use the upstream version of the kubelet which will result in Red Hat being
+responsible for security and other fixes.
+
+The other risk is "OVN Plumbing for Hybrid Linux+Windows Clusters GA" not being
+delivered in time for integration. The mitigation for that would be to use 3rd
+party networking components but that might have support implications that would
+need to be resolved.
+
+## Design Details
+
+### Test Plan
+
+We plan to have all the repositories associated with this effort fully
+integrated with Prow CI and run e2e tests for every PR that is opened. These
+e2e tests will involve bringing up a cluster on all supported cloud providers,
+instantiating a Windows node and running workloads on it. We also plan to add
+blocking tests to the nightly runs.
+
+### Graduation Criteria
+
+This enhancement will start as GA
+
+### Upgrade / Downgrade Strategy
+
+We will support upgrades of the node components by publishing a new release
+of the WSU Ansible playbook. An older release of the playbook can be used to
+downgrade.
+
+### Version Skew Strategy
+
+We plan to maintain kubelet major version parity with the Linux counterpart.
+
+## Implementation History
+
+v1: Initial proposal
+
+## Drawbacks
+
+The Microsoft container ecosystem is not fully mature or on par with Linux
+containers. For one we can only support Windows process containers that
+restricts workloads that are runnable on Windows Server 2019 as there is a tight
+kernel version coupling with Windows containers. The coupling is slightly
+loosened if we use Hyper-V containers but that is not a viable option yet with
+all cloud providers as it requires nested virtualization support. Given these
+limitations customers could potentially get a degraded experience with Windows
+workloads when compared to Linux workloads on the cluster.
+
+## Alternatives
+
+An alternate approach to this design is to follow the operator approach where in
+the user will first install an operator from OperatorHub. The operator will
+introduce a CRD that will take similar inputs that the Ansible playbook does.
+It will then ensure that the WMCB is installed and launched on the cluster. The
+WMCB could also potentially become Windows Machine Config Daemon (WMCD) that
+supports a run once option. While it is a daemon it would run as a Windows
+service constantly reconciling with a config object on the cluster to ensure the
+node is in the desired state.
+
+We opted not to go with this approach given then time frame all the unknowns
+present in this project.
```

Comment:
Ahh, some of my previous comments were made under the assumption we'd still have an operator distributed via OLM. So this proposal now is completely external to the cluster just like BYOH RHEL playbooks are. I think this is a fine place to start but lets consider a future that uses OLM to distribute things like the WMCB, if we end up sticking with Ansible for remote management we could move the playbooks from admin initiated to an operator that runs ansible inside a pod.

Reply:
We are definitely going the OLM operator route moving forward. This is also the reason to to have WSU - WMCB split. The WSU could very well become an operator and WMCB would become a daemon (Windows service) running on the node.